src/smolagents/agents.py

```diff
@@ -1258,18 +1275,44 @@ def step(self, memory_step: ActionStep) -> Union[None, Any]:
         memory_messages = self.write_memory_to_messages()
 
         self.input_messages = memory_messages.copy()
-
         # Add new step in logs
         memory_step.model_input_messages = memory_messages.copy()
         try:
             additional_args = {"grammar": self.grammar} if self.grammar is not None else {}
-            chat_message: ChatMessage = self.model(
-                self.input_messages,
-                stop_sequences=["<end_code>", "Observation:", "Calling tools:"],
-                **additional_args,
-            )
-            memory_step.model_output_message = chat_message
-            model_output = chat_message.content
+            if self.stream_outputs:
+                output_stream = self.model.generate_stream(
+                    self.input_messages,
+                    stop_sequences=["<end_code>", "Observation:", "Calling tools:"],
+                    **additional_args,
+                )
+                output_text = ""
+                with Live("", console=self.logger.console, vertical_overflow="visible") as live:
+                    for event in output_stream:
+                        memory_step.model_output_message = event
+                        if isinstance(event, CompletionDelta):
+                            if event.content is not None:
+                                output_text += event.content
+                                live.update(Markdown(output_text))
+                        elif isinstance(event, ToolCall):
+                            memory_step.tool_calls = [event]
+
+                model_output = output_text
+                chat_message = ChatMessage(role="assistant", content=model_output)
+                memory_step.model_output_message = chat_message
+                model_output = chat_message.content
+            else:
+                chat_message: ChatMessage = self.model.generate(
```

Comment:
To maintain backward compatibility:
```suggestion
                chat_message: ChatMessage = self.model(
```